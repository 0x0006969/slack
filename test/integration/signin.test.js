const request = require('supertest');
const nock = require('nock');

const helper = require('.');
const fixtures = require('../fixtures');

const { probot } = helper;

describe('Integration: signin', () => {
  describe('unauthenticated user', () => {
    test('is prompted to authenticate', async () => {
      // User types slash command
      const command = fixtures.slack.command({
        text: 'signin',
      });
      const req = request(probot.server).post('/slack/command').send(command);
      const res = await req.expect(200);

      // User is shown ephemeral prompt to authenticate
      const prompt = /^<https:\/\/example\.com(\/github\/oauth\/login\?state=(.*))\|Finish connecting your GitHub account>$/;
      const text = res.body.attachments[0].text;
      expect(text).toMatch(prompt);

      // User follows link to OAuth
      const [, link, state] = text.match(prompt);

      const loginRequest = request(probot.server).get(link);
      await loginRequest.expect(302).expect('Location',
        `https://github.com/login/oauth/authorize?client_id=&state=${state}`,
      );

      // GitHub redirects back, authenticates user and process subscription
      nock('https://github.com').post('/login/oauth/access_token')
        .reply(200, fixtures.github.oauth);
      nock('https://api.github.com').get('/user')
        .reply(200, fixtures.user);

      nock('https://hooks.slack.com').post('/commands/1234/5678', {
        response_type: 'ephemeral',
        attachments: [{
          text: `:white_check_mark: Success! <@${command.user_id}> is now connected to <${fixtures.user.html_url}|@${fixtures.user.login}>`,
        }],
      }).reply(200);

      await request(probot.server).get('/github/oauth/callback').query({ state })
        .expect(302)
        .expect('Location',
          `slack://channel?team=${command.team_id}&channel=${command.channel_id}`,
        );
    });
  });
});
