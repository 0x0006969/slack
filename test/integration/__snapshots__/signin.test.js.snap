// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: signin unauthenticated user is prompted to authenticate 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "mrkdwn_in": Array [
        "text",
      ],
      "text": ":white_check_mark: Success! <@U2147483697> is now connected to <https://github.com/wilhelmklopp|@wilhelmklopp>",
    },
  ],
  "response_type": "ephemeral",
}
`;

exports[`Integration: signin with a pending subscription redirects to install app and creates subscription 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "text": "Subscribed to <https://github.com/bkeepers/dotenv|bkeepers/dotenv>",
    },
  ],
  "response_type": "in_channel",
}
`;

exports[`Integration: signin with a pending subscription redirects to install app and creates subscription 2`] = `
[MockFunction cache.get] {
  "calls": Array [
    Array [
      "pending-command#13345224609.738474920.8088930838d88f008e0",
    ],
    Array [
      "pending-command#13345224609.738474920.8088930838d88f008e0",
    ],
    Array [
      "pending-command#13345224609.738474920.8088930838d88f008e0",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Integration: signin with a pending subscription redirects to install app and creates subscription 3`] = `
[MockFunction cache.set] {
  "calls": Array [
    Array [
      "pending-command#13345224609.738474920.8088930838d88f008e0",
      Object {
        "channel_id": "C2147483705",
        "channel_name": "test",
        "command": "/github",
        "enterprise_id": "E0001",
        "enterprise_name": "Globular%20Construct%20Inc",
        "response_url": "https://hooks.slack.com/commands/1234/5678",
        "team_domain": "example",
        "team_id": "T0001",
        "text": "subscribe kubernetes/kubernetes",
        "token": "secret",
        "trigger_id": "13345224609.738474920.8088930838d88f008e0",
        "user_id": "U2147483697",
        "user_name": "Steve",
      },
    ],
    Array [
      "pending-command#13345224609.738474920.8088930838d88f008e0",
      Object {
        "channel_id": "C2147483705",
        "channel_name": "test",
        "command": "/github",
        "enterprise_id": "E0001",
        "enterprise_name": "Globular%20Construct%20Inc",
        "response_url": "https://hooks.slack.com/commands/1234/5678",
        "team_domain": "example",
        "team_id": "T0001",
        "text": "subscribe kubernetes/kubernetes",
        "token": "secret",
        "trigger_id": "13345224609.738474920.8088930838d88f008e0",
        "user_id": "U2147483697",
        "user_name": "Steve",
      },
    ],
    Array [
      "pending-command#13345224609.738474920.8088930838d88f008e0",
      Object {
        "channel_id": "C2147483705",
        "channel_name": "test",
        "command": "/github",
        "enterprise_id": "E0001",
        "enterprise_name": "Globular%20Construct%20Inc",
        "response_url": "https://hooks.slack.com/commands/1234/5678",
        "team_domain": "example",
        "team_id": "T0001",
        "text": "subscribe kubernetes/kubernetes",
        "token": "secret",
        "trigger_id": "13345224609.738474920.8088930838d88f008e0",
        "user_id": "U2147483697",
        "user_name": "Steve",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Integration: signin with a pending subscription redirects to install app and creates subscription 4`] = `[MockFunction cache.fetch]`;
